### PR TITLE
Polymorphic refinements

### PR DESCRIPTION
**Pull request by lewisjkl**
Adds the ability to have refinements which abstract over a type parameter (or multiple parameters in the case of Map). See the added docs in the PR for more information.

### CODE REVIEW

File: modules/docs/markdown/04-codegen/01-customisation.md [hunk 1]
```diff
@@ -270,6 +270,50 @@ Whether the provider is in the companion object or not, it must be `implicit`.
 
 :::
 
+#### Parameterised Types
+
+As of smithy4s version 0.17, you can now create refinements on types that take a generic type parameter. This can be accomplished by setting `parameterised` to `true` as seen below.
+
+```kotlin
```
Comment (kubukoz) [inline modules/docs/markdown/04-codegen/01-customisation.md:277]:
```suggestion
```smithy
```

File: modules/codegen/src/smithy4s/codegen/ToLine.scala [hunk 2]
```diff
@@ -67,12 +67,26 @@ object ToLine {
...
+          if (acc.nonEmpty) acc + Literal(",") + Line.space + i else i
+      }
+    val nameLine = NameRef(ext.fullyQualifiedName).toLine
+    val result =
+      if (paramLines.nonEmpty)
+        nameLine + Literal("[") + paramLines + Literal("]")
+      else nameLine
+    result
+  }
+
```
Comment (lewisjkl) [inline modules/codegen/src/smithy4s/codegen/ToLine.scala:89]: I think I can improve on this by updating NameRef to instead take in the type parameter information. Going to give it a go at least

File: modules/docs/markdown/04-codegen/01-customisation.md (hunk 1 above)
Comment (lewisjkl) [reply modules/docs/markdown/04-codegen/01-customisation.md:277]: Not sure if our docusaurus site does highlighting for smithy yet

File: modules/codegen/src/smithy4s/codegen/ToLine.scala (hunk 2 above)
Comment (lewisjkl) [reply modules/codegen/src/smithy4s/codegen/ToLine.scala]: Done in https://github.com/disneystreaming/smithy4s/pull/649/commits/0fb6a53717fec06ccf4a1d7d8a6fb861cd5faef3

**Review by Baccata — APPROVED** [review]

**Pull request merged**